core/state_transition.go

```diff
@@ -163,12 +165,64 @@ func IntrinsicGas(data []byte, accessList types.AccessList, isContractCreation b
 		gas += z * params.TxDataZeroGas
 	}
 	if accessList != nil {
+		accessListGas, err := AccessListGas(rules, accessList)
+		if err != nil {
+			return 0, err
+		}
+		totalGas, overflow := commonMath.SafeAdd(gas, accessListGas)
+		if overflow {
+			return 0, fmt.Errorf("overflow adding access list gas")
+		}
+		gas = totalGas
+	}
+
+	return gas, nil
+}
+
+func AccessListGas(rules params.Rules, accessList types.AccessList) (uint64, error) {
+	var gas uint64
+	if !rules.PredicatesExist() {
 		gas += uint64(len(accessList)) * params.TxAccessListAddressGas
 		gas += uint64(accessList.StorageKeys()) * params.TxAccessListStorageKeyGas
+		return gas, nil
 	}
+
+	for i, accessTuple := range accessList {
+		address := accessTuple.Address
+		if !rules.PredicateExists(address) {
+			totalGas, overflow := commonMath.SafeAdd(gas, params.TxAccessListAddressGas+uint64(len(accessTuple.StorageKeys))*params.TxAccessListStorageKeyGas)
+			if overflow {
+				return 0, fmt.Errorf("overflow adding storage slot gas at index %d", i)
+			}
```

Comment:
technically this does represent a behavior change. Since it's enabled via a network upgrade it should be fine, but still perhaps we should retain old behavior? (is it even possible to overflow here?)

Reply:
With the current code and gas charges for specifying state storage slot tuples, it would take multiple TBs of data in order for an access list to overflow uint64.

This only changes the behavior when a predicate could return an arbitrary amount of gas, such that an overflow becomes realistic here.

I'm going to change this to multiple lines and keep the safemath when adding to the total since a precompile predicate gas implementation can return a `uint64` that could result in an overflow, so we should check that case here.